### PR TITLE
Rework environment handling and unification for cycle detection

### PR DESCRIPTION
**Pull request by mistasse**
Basically to complete the integration of the changes from my master's thesis!

### CODE REVIEW

File: bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala [hunk 1]
```diff
@@ -30,7 +30,7 @@ sealed class VarState(val onStack: Boolean, val locked: Boolean) {
   }
 }
 object Declared      extends VarState(true, false)
-object Captured      extends VarState(false, false)
+object Captured      extends VarState(false, true) // TODO false false if frame hierarchy, false true if extraction
```
Comment (mistasse) [inline bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala:33]: It would be great to parse the params

**Comment by eregon** [comment]
Commits until 5d48e83 are in #19, can we merge that one first to avoid mixing too many changes? :smiley:

**Comment by mistasse** [comment]
Definitely!

**Comment by eregon** [comment]
Thanks! I merged the changes to master.

**Pull request merged**